EngineAPI/src/main/java/nullengine/event/world/block/BlockInteractEvent.java

```diff
@@ -0,0 +1,92 @@
+package nullengine.event.world.block;
+
+import org.joml.Vector3fc;
+
+import nullengine.block.Block;
+import nullengine.event.Cancellable;
+import nullengine.event.Event;
+import nullengine.event.world.block.cause.BlockInteractCause;
+import nullengine.math.BlockPos;
+import nullengine.util.Facing;
+import nullengine.world.World;
+
+public abstract class BlockInteractEvent implements Event, Cancellable {
+    private boolean isCancelled;
+
+    private World world;
+    private Block block;
+    /**
+     * World space block
+     */
+    private BlockPos position;
+    /**
+     * The hit vector should be in block local space
+     */
+    private Vector3fc hit;
+
+    private Facing face;
+
+    /**
+     * The cause contains the reason & the type of the interact.
+     */
+    private BlockInteractCause cause;

```

Comment:
In this design, dev will determine whether this is a hit (left click) or a use (right click) by this field.

See entity trigger